File: credit.c
#include <cs50.h>
#include <stdio.h>

long card;
int sum1=0;
int sum2=0;
int length=0;
int checksum=0;
int main(void)
{
    card = get_long("Please enter card number\n");
    /* checksum */
    for( long i = card ; i > 0; i=i/100){
        sum1 = sum1 + i%10;
    }
    for( long j = card*10 ; j > 0; j=j/100){
        sum2 = sum2 + j%10;
    }
    checksum = sum1 + sum2;    
    if(checksum%10 != 0){
        printf("INVALID");
    }
    else
    {
        for(long z = card; z > 0 ; z=z/10){
        length = length + 1;
    }
    if((length == 16 && card/1000000000000000%10 == 5)&&(card/100000000000000%10 >0 && card/100000000000000%10 <6)){
        printf("MasterCard");
    }
    else{
        if((length == 15 && card/100000000000000%10 == 3)&&(card/10000000000000%10 == 4 || card/10000000000000%10 == 7 )){
            printf("American Express");
        }
        else{
        if((length == 13 && card/1000000000000%10 == 4)||(length == 16 && card/1000000000000000%10 == 4)){
            printf("VISA");
        }
         else{
            printf("INVALID");     
            }
        }
    }
}
}

File: credit.py
from cs50 import get_int
from cs50 import get_string
import sys
def main():
    card = get_card()
    validate(card)

def get_card():
    credit = get_int("Number: ")
    length = len(str(credit))
    if length != 13 and length != 15 and length != 16:
        print("INVALID\n")
        sys.exit()
    else:
        return credit

def validate(card):
    length = len(str(card))
    dig1 = 0
    dig2 = 0
    sum1 = 0
    sum2 = 0
    digcount = 0
    while card > 0:
        dig2 = dig1
        dig1 = card % 10
        if digcount % 2 == 0:
            sum1 += dig1
        else:
            sum2 = sum2 + ((dig1*2)//10) + ((dig1*2)%10)
        card = card // 10
        digcount += 1
    validtest = (sum1 + sum2) % 10
    fdig = (dig1*10) + dig2
    if length >= 13 and length <=16 and dig1 ==4 and validtest == 0:
        print("VISA\n")
    elif fdig <=55 and fdig >= 51 and length == 16 and validtest ==0:
        print("MASTERCARD\n")
    elif length == 15 and validtest == 0 and (fdig == 34 or fdig == 37):
        print("AMEX\n")
    else:
        print("INVALID\n")

main()